File: MATLAB/NonlinearRegression.m
% Read the CSV file
data = csvread('Walmart_Sales.csv', 1, 0);

% Extract columns
x = data(:, 8);  % Unemployment
y = data(:, 3);  % Weekly_sales

% Define the nonlinear model function as y = a * (1 - e^(b * x)) + c
function y_model = model(params, x)
    a = params(1);
    b = params(2);
    c = params(3);
    y_model = a * (1 - exp(b * x)) + c;
end

% Define function for SSE
function sse = SSE(params, x, y)
    y_model = model(params, x);
    sse = sum((y - y_model) .^ 2);
end

% Initial guess for a, b, c
initial_params = [1.0, -0.1, 1.0];

% Search the optimal a, b, and c using fminsearch function
result = fminsearch(@(params) SSE(params, x, y), initial_params);

% Get the optimized parameters
a_opt = result(1);
b_opt = result(2);
c_opt = result(3);

% Print final parameters
printf('Optimized parameters: a = %.4f, b = %.4f, c = %.4f\n', a_opt, b_opt, c_opt);

% Calculate fitted values
y_pred = model(result, x);

% Create the plot
figure;
scatter(x, y, 'b', 'DisplayName', 'Data points');
hold on;
plot(x, y_pred, 'r', 'DisplayName', sprintf('Nonlinear regression: y = %.2f * (1 - e^{%.2f * x}) + %.2f', a_opt, b_opt, c_opt));

% Add title & label
title('Weekly Sales vs. Unemployment with Nonlinear Regression');
xlabel('Unemployment');
ylabel('Weekly Sales');
legend('show');
set(gca, 'FontSize', 24);

File: NonlinearRegression.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.optimize

# Read csv file
df = pd.read_csv('Walmart_Sales.csv')

# Get unemployment and weekly_sales data
x = df['Unemployment'].values
y = df['Weekly_Sales'].values

# Define the nonlinear model function as y = a * (1 - e^(b * x)) + c
def model(params, x):
    a, b, c = params
    return a * (1 - np.exp(b * x)) + c

# Define function for SSE
def SSE(params, x, y):
    y_model = model(params, x)
    return np.sum((y - y_model) ** 2)

# Initial guess for a, b, c
initial_params = [1.0, -0.1, 1.0]

# Search the optimal a, b, and c using fmin function
result = scipy.optimize.fmin(func=SSE, x0=initial_params, args=(x, y), xtol=1e-8, ftol=1e-8)

# Get the optimized parameters
a_opt, b_opt, c_opt = result

# Print final parameters
print(f'Optimized parameters: a = {a_opt:.4f}, b = {b_opt:.4f}, c = {c_opt:.4f}')

# Calculate fitted values
y_pred = model(result, x)

# Create the plot
plt.figure(figsize=(10, 6))
plt.scatter(x, y, color='blue', label='Data points')
plt.plot(x, y_pred, color='red', label=f'Nonlinear regression: y = {a_opt:.2f} * (1 - e^({b_opt:.2f} * x)) + {c_opt:.2f}')

# Add title & label
plt.title('Weekly Sales vs. Unemployment with Nonlinear Regression')
plt.xlabel('Unemployment')
plt.ylabel('Weekly Sales')
plt.legend()

# Show the plot
plt.grid(True)
plt.show()
